web-api/terraform/modules/batch/docker-image/src/docketEntryBatchDownloader.ts

```diff
@@ -0,0 +1,263 @@
+}
+
+async function getDocketEntriesFromS3(
+  documentsReference: string,
+  s3Client: S3,
+) {
+  const fetchDocumentsReferenceCommand = new GetObjectCommand({
+    Bucket: TEMP_S3_BUCKET,
+    Key: documentsReference,
+  });
+
+  const documentsReferenceData = await s3Client.send(
+    fetchDocumentsReferenceCommand,
+  );
+
+  if (!documentsReferenceData.Body) throw new Error('Unable to get documents');
+
+  const docketEntriesBuffer: any = await streamToBuffer(
+    documentsReferenceData.Body,
+  );
+  const docketEntriesString: string = docketEntriesBuffer.toString('utf-8');
+  const docketEntries: DocketEntryDownloadInfo[] =
+    JSON.parse(docketEntriesString);
+  return docketEntries;
+}
+
+export async function zipDocuments({
+  connectionId,
+  documents,
+  outputZipName,
+  wsClient,
+}: {
+  wsClient: any;
+  connectionId: string;
+  documents: {
+    key: string;
+    filePathInZip: string;
+    useTempBucket: boolean;
+  }[];
+  outputZipName: string;
+}): Promise<void> {
+  const passThrough = new PassThrough({ highWaterMark: 1024 * 1024 * 100 });
+
+  const upload = new Upload({
+    client: zipStorageClient,
+    params: {
+      Body: passThrough,
+      Bucket: ZIP_TEMP_S3_BUCKET,
+      Key: outputZipName,
+    },
+  });
+
+  const writable = new Writable({
+    write(chunk, encoding, callback) {
+      callback();
+    },
+  });
+
+  passThrough.pipe(writable);
+
+  // Start the upload process
+  const uploadPromise = upload.done();
+
+  const zip = new Zip((err, data, final) => {
+    if (err) {
+      console.log('Error creating zip stream');
+      throw err;
+    }
+
+    passThrough.write(data);
+
+    if (final) {
+      passThrough.end();
+    }
+  });
+
+  for (let index = 0; index < documents.length; index++) {
+    const { filePathInZip, key, useTempBucket } = documents[index];
+    console.log(`Downloding document (${key})`);
+
+    const response = await storageClient.getObject({
+      Bucket: useTempBucket ? TEMP_S3_BUCKET : S3_BUCKET,
+      Key: key,
+    });
+
+    if (!response.Body) {
+      throw new Error(`Unable to get document (${key}) from persistence.`);
+    }
+
+    // Transform s3 getobject into a stream of data that can be piped into the zip processor
+    const bodyStream: ReadableStream<Uint8Array> =
+      response.Body.transformToWebStream();
+    const reader = bodyStream.getReader();
+    const compressedPdfStream = new AsyncZipDeflate(filePathInZip);
+    zip.add(compressedPdfStream);
+
+    let continueReading = true;
+    while (continueReading) {
+      const unzippedChunk = await reader.read();
+      const nextChunk = unzippedChunk.value || new Uint8Array();
+      continueReading = !unzippedChunk.done;
+      compressedPdfStream.push(nextChunk, unzippedChunk.done);
+      while (passThrough.readableLength > 1024 * 1024 * 10) {
+        // Wait for the buffer to be drained, before downloading more files
+        await new Promise(resolve => setTimeout(resolve, 10));
+      }
+    }
```

Comment:
Is it possible to use [pipeline](https://nodejs.org/api/stream.html#streampipelinesource-transforms-destination-options) here? E.g., something like `await pipeline(bodyStream, compressedPdfStream, passThrough);`